lib/proxy/auth.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/credentials"
+)
+
+// newProxyCredentials creates new proxyCredentials from the given transport credentials.
+func newProxyCredentials(creds credentials.TransportCredentials) credentials.TransportCredentials {
+	return &proxyCredentials{
+		creds,
+	}
+}
+
+// proxyCredentials wraps TransportCredentials server and client handshakes
+// to ensure the credentials contain the proxy system role.
+type proxyCredentials struct {
+	credentials.TransportCredentials
+}
+
+// ServerHandshake wraps a server handshake with an additional check for the
+// proxy role.
+func (c *proxyCredentials) ServerHandshake(conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, authInfo, err := c.TransportCredentials.ServerHandshake(conn)
+	if err != nil {
+		return conn, authInfo, err
+	}
+
+	err = checkProxyRole(authInfo)
+	return conn, authInfo, trace.Wrap(err)
```

Comment:
Is it intended that `conn` and `authInfo` can be non-nil with a non-nil error?

Reply:
Fixed here [39555e5](https://github.com/gravitational/teleport/pull/10223/commits/39555e5f1decac8ab2794dee733ac936c6a7b083)